lib/src/revset.rs

```diff
@@ -1744,6 +1790,101 @@ fn fold_ancestors_union<St: ExpressionState>(
     })
 }
 
+/// Transforms expressions like `heads(roots..heads & filters)` into a combined
+/// operation where possible. Ancestors and negated ancestors should have
+/// already been moved to the left in intersections, and negated ancestors
+/// should have been combined already.
+fn fold_heads_range<St: ExpressionState>(
+    expression: &Rc<RevsetExpression<St>>,
+) -> TransformedExpression<St> {
+    // Represents `roots..heads & filter`
+    struct FilteredRange<St: ExpressionState> {
+        roots: Rc<RevsetExpression<St>>,
+        heads: Option<Rc<RevsetExpression<St>>>,
+        filter: Rc<RevsetExpression<St>>,
+    }
+
+    impl<St: ExpressionState> FilteredRange<St> {
+        fn new(roots: Rc<RevsetExpression<St>>) -> Self {
+            // roots.. & all()
+            Self {
+                roots,
+                heads: None,
+                filter: RevsetExpression::all(),
+            }
+        }
+
+        fn add(mut self, expression: &Rc<RevsetExpression<St>>) -> Self {
+            if self.heads.is_none() {
+                // x.. & ::y -> x..y
+                if let Ok(heads) = ancestors_to_heads(expression) {
+                    self.heads = Some(heads);
+                    return self;
+                }
+            }
+
+            if let RevsetExpression::All = self.filter.as_ref() {
```

Comment:
nit: I slightly prefer splitting this function here, and call `.add_filter()` where `expression` is obviously not `::heads`.